File: aoc2024/01/main.py
from functools import reduce

FILE = "data.txt"
#FILE = "test.txt"

def parse_line(line):
    return list(map(int, line.split(" " * 3)))

def load_data():
    list1, list2 = [], []
    with open(FILE, "r") as f:
        for line in f:
            d = parse_line(line)
            list1.append(d[0])
            list2.append(d[1])
    return list1, list2

def sol1():
    list1, list2 = load_data()
    res = reduce(lambda acc, t: acc + abs(t[0] - t[1]), zip(sorted(list1), sorted(list2)), 0)
    print("Solution 1:", res) 

def sol2():
    list1, list2 = load_data()
    res = reduce(lambda acc, x: acc + x * list2.count(x), list1, 0)
    print("Solution 2:", res)
    
if __name__ == "__main__":
    sol1()
    sol2()

File: aoc2024/01/main.hs
import Data.List

filename = "data.txt"
-- filename = "test.txt"

parse_line :: String -> [Int]
parse_line line = map read $ words line

sol1 lines = foldl (\acc (x, y) -> acc + abs (x - y)) 0 (zip (sort first_lst) (sort second_lst))
  where
    first_lst = map (head . parse_line) $ lines 
    second_lst = map (head . tail . parse_line) $ lines

count :: Int -> Int -> [Int] -> Int
count needle acc [] = acc
count needle acc (x:xs)
  | x == needle = count needle (acc + 1) xs
  | otherwise = count needle acc xs

sol2 lines = foldl (\acc x -> acc + x * (count x 0 second_lst)) 0 first_lst 
  where
    first_lst = map (head . parse_line) $ lines 
    second_lst = map (head . tail . parse_line) $ lines

main :: IO ()
main = do
  contents <- readFile filename
  putStrLn $ "Solution 1: " ++ ((show . sol1 . lines) contents)
  putStrLn $ "Solution 2: " ++ ((show . sol2 . lines) contents)
  return ()
  
  
